exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchClient.java

```diff
@@ -0,0 +1,225 @@
+import java.time.format.DateTimeFormatter;
+import java.util.Collections;
+import java.util.Map;
+import org.apache.http.HttpHost;
+import org.apache.http.impl.nio.reactor.IOReactorConfig;
+import org.elasticsearch.action.admin.indices.template.put.PutIndexTemplateRequest;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestClientBuilder;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.slf4j.Logger;
+
+public class ElasticsearchClient {
+
+  public static final String INDEX_TEMPLATE_FILENAME_PATTERN = "/zeebe-record-%s-template.json";
+
+  private final ElasticsearchExporterConfiguration configuration;
+  private final Logger log;
+  private final RestHighLevelClient client;
+  private BulkRequest bulkRequest;
+
+  private final DateTimeFormatter formatter;
+
+  public ElasticsearchClient(final ElasticsearchExporterConfiguration configuration, Logger log) {
+    this.configuration = configuration;
+    this.log = log;
+    this.client = createClient();
+    this.bulkRequest = new BulkRequest();
+    this.formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd").withZone(ZoneOffset.UTC);
+  }
+
+  public void close() throws IOException {
+    client.close();
+  }
+
+  public void index(final Record<?> record) {
+    final IndexRequest request =
+        new IndexRequest(indexFor(record), typeFor(record), idFor(record))
+            .source(record.toJson(), XContentType.JSON);
+    bulk(request);
+  }
+
+  public void bulk(final IndexRequest indexRequest) {
+    bulkRequest.add(indexRequest);
+  }
+
+  /** @return true if all bulk records where flushed successfully */
+  public boolean flush() {
+    boolean success = true;
+    if (bulkRequest.numberOfActions() > 0) {
+      try {
+        final BulkResponse responses = client.bulk(bulkRequest, RequestOptions.DEFAULT);
+        success = checkBulkResponses(responses);
+      } catch (IOException e) {
+        throw new ElasticsearchExporterException("Failed to flush bulk", e);
+      }
+
+      if (success) {
+        // all records where flushed, create new bulk request, otherwise retry next time
+        bulkRequest = new BulkRequest();
+      }
+    }
+
+    return success;
+  }
+
+  private boolean checkBulkResponses(final BulkResponse responses) {
+    boolean failed = false;
+
+    for (BulkItemResponse response : responses) {
+      if (response.isFailed()) {
+        failed = true;
+        log.warn("Failed to flush bulk request {}", response.getFailureMessage());
```

Comment:
Can exit early on failure; is it important to see all failures?

Reply:
I wasn't sure, I first had the early return but thought it might be interesting to know how many failed because then you basically have some missing records in elastic as the bulk is not *atomic* or *transactional*. I don't have a strong opinion on that and can change it if you want.